### ISSUE TITLE
Allow Collations to be Swithed During Upgrade

### ISSUE THREAD

**Issue by interduo** [Cacti/cacti#1034]
```
2017/10/16  08:59:55 - CMDPHP SQL Backtrace: (/host.php: 136 host)(/host.php: 1615  get_device_records)(/host.php: 1352 db_fetch_assoc)(/lib/database.php:  360 db_fetch_assoc_prepared)(/lib/database.php: 401  cacti_debug_backtrace)
--
2017/10/16 08:59:55 - DBCALL ERROR: SQL Assoc Failed!, Error:  Illegal mix of collations (latin1_swedish_ci,IMPLICIT) and  (utf8_general_ci,COERCIBLE) for operation 'like'
2017/10/16 08:59:55 - DBCALL ERROR: SQL Assoc Failed!, Error:1267,  SQL:'SELECT host.*, graphs, data_sources,  IF(status_event_count>0,  status_event_count*300,    IF(UNIX_TIMESTAMP(status_rec_date)>943916400,UNIX_TIMESTAMP()-UNIX_TIMESTAMP(status_rec_date),    IF(snmp_sysUptimeInstance>0 AND snmp_version > 0,  snmp_sysUptimeInstance,UNIX_TIMESTAMP()))) AS instate  FROM host  LEFT  JOIN (SELECT host_id, COUNT(*) AS graphs FROM graph_local GROUP BY  host_id) AS gl  ON host.id=gl.host_id  LEFT JOIN (SELECT host_id,  COUNT(*) AS data_sources FROM data_local GROUP BY host_id) AS dl  ON  host.id=dl.host_id  WHERE (host.hostname LIKE '%Niedźwiada%' OR  host.description LIKE '%Niedźwiada%')  GROUP BY host.id  ORDER BY  `description` ASC  LIMIT 0,1000'
```

If You put Polish letters 'żćółń' - there is SQL error, but no filtering device list.

**Comment by interduo** [comment]
If You put polish żćółń in SQL there is an error, but no filtering this.

**Comment by cigamit** [comment]
What is your DB's Collation?  Everything needs to be UTF-8 (mb4).  You likely either have columns or entire tables in latin_swedish_ci.  You can see if you dump the database on the table create lines.

**Comment by interduo** [comment]
Is it possible to change collation without stoping engine?

**Comment by cigamit** [comment]
You have to use alter statements on the tables affected, yes, you can do this without a restart.  However, you better do a dump just in case something get's screwed up.  You can dump the tables individually, or the entire database.  Your choice.  Use google to guide you on this one.

**Comment by interduo** [comment]
#!/bin/bash
USER="pleasetypeyourlogin";
PASSWORD="pleasetypeyourpassword";
DB_NAME="cacti"
CHARACTER_SET="utf8mb4" # your default character set
COLLATE="utf8mb4_unicode_ci" # your default collation

tables=`mysql -u $USER -p$PASSWORD -e "SELECT tbl.TABLE_NAME FROM information_schema.TABLES tbl WHERE tbl.TABLE_SCHEMA = '$DB_NAME' AND tbl.TABLE_TYPE='BASE TABLE'"`

for tableName in $tables; do
if [[ "$tableName" != "TABLE_NAME" ]] ; then
mysql -u $USER -p$PASSWORD -e "ALTER TABLE $DB_NAME.$tableName CONVERT TO CHARACTER SET $CHARACTER_SET COLLATE $COLLATE;"

echo "$tableName - done"
fi
done

**Comment by interduo** [comment]
You were right. Maybe installer should check collation when upgrading?

**Comment by cigamit** [comment]
Yea, good point, might be pretty easy to accomplish.

**Comment by tosiara** [comment]
I second this enhancement. The provided script helped me, thanks

**Comment by netniV** [comment]
This has been added as a 1.2 feature request in #1482

**Comment by netniV** [comment]
This has now been implemented in the 1.2 branch as part of the installer for any table with less than 300,000 rows.
